File: bandlog-team/src/main/java/net/effize/bandlog/team/service/TeamService.java
package net.effize.bandlog.team.service;

import net.effize.bandlog.team.model.Member;
import net.effize.bandlog.team.model.MemberRole;
import net.effize.bandlog.team.model.Team;
import net.effize.bandlog.team.model.TeamId;
import net.effize.bandlog.team.model.UserId;
import net.effize.bandlog.team.repository.MemberRepository;
import net.effize.bandlog.team.repository.TeamRepository;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;

@Service
public class TeamService {
    private final TeamRepository teamRepository;
    private final MemberRepository memberRepository;

    public TeamService(TeamRepository teamRepository, MemberRepository memberRepository) {
        this.teamRepository = teamRepository;
        this.memberRepository = memberRepository;
    }

    public Team createTeam(String name, String description, Instant now) {
        Team newTeam = Team.create(name, description, now);
        return teamRepository.save(newTeam);
    }

    public Team addNewMember(Team team, UserId userId, MemberRole role, Instant now) {
        Member newMember = Member.create(team, userId, role, now);
        memberRepository.save(newMember);
        return team;
    }

    public Team activeTeam(TeamId id) {
        return teamRepository.findById(id.value())
                .orElseThrow();
    }

    public List<Member> membersOf(Team team) {
        return memberRepository.findAllByTeam(team);
    }

    public List<Team> teamsOfUser(UserId userId) {
        return teamRepository.findAllByMembersUserId(userId.value());
    }
}

File: src/main/kotlin/net/effize/bandlog/team/application/TeamService.kt
package net.effize.bandlog.team.application

import net.effize.bandlog.common.id.UserId
import net.effize.bandlog.team.adapter.out.persistence.MemberJpaEntity
import net.effize.bandlog.team.adapter.out.persistence.MemberJpaRepository
import net.effize.bandlog.team.adapter.out.persistence.TeamJpaEntity
import net.effize.bandlog.team.adapter.out.persistence.TeamJpaRepository
import net.effize.bandlog.team.domain.MemberId
import net.effize.bandlog.team.domain.MemberRole
import net.effize.bandlog.team.domain.TeamId
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class TeamService(
    private val teamRepository: TeamJpaRepository,
    private val memberRepository: MemberJpaRepository
) {

    fun createTeam(name: String, description: String, now: Instant): TeamJpaEntity {
        val newTeam = TeamJpaEntity.create(name, description, now)
        return teamRepository.save(newTeam)
    }

    fun addNewMember(team: TeamJpaEntity, userId: UserId, role: MemberRole, now: Instant): TeamJpaEntity {
        val newMember = MemberJpaEntity.create(team, userId, role, now)
        memberRepository.save(newMember)
        return team
    }

    fun activeTeam(id: TeamId): TeamJpaEntity {
        return teamRepository.findById(id.value).orElseThrow()
    }

    fun membersOf(team: TeamJpaEntity): List<MemberJpaEntity> {
        return memberRepository.findAllByTeam(team)
    }

    fun teamsOfUser(userId: UserId): List<TeamJpaEntity> {
        return teamRepository.findAllByMembersUserId(userId.value)
    }

    fun memberOf(id: MemberId): MemberJpaEntity {
        return memberRepository.findById(id.value).orElseThrow()
    }
}
